### PR TITLE
Heaviside: use None to specify default 2nd-argument behaviour

### PR DESCRIPTION
**Pull request by cbm755**
This is on top of #11065

### CODE REVIEW

**Comment by cbm755** [comment]
@zanzibar7 What do you think?

The potentially controversial bit is e5b3103...

**Comment by cbm755** [comment]
I don't want the `None` to survive object creation (hence test just added).

I guess the following is undesirable but acceptable:
````
In [4]: Heaviside(x, None, evaluate=False)
Out[4]: Heaviside(x, None)
````

**Comment by jksuom** [comment]
I would prefer using `None` instead of `NaN`. It can be interpreted as saying that `H(0)` is undefined, which I think is a reasonable choice. `NaN` would seem to imply that the value is defined but unknown.

**Comment by cbm755** [comment]
Ok, do you also have a preference for implementation between this way (sort-of allowing None in args, at least temporarily) versus #11244 (writing our own `__new__`)?

**Comment by jksuom** [comment]
My preference is on the alternative implementation with `Heaviside.__new__` rewritten. Then it is not necessary to modify the way functions are evaluated, all the changes are confined to `Heaviside` (which, I think, should be considered as a 'generalized function').

**Comment by zanzibar7** [comment]
`NaN` occurs naturally in the kinds of calculations associated with `Heaviside(0)`.  In particular, situations where direct evaluation gives `0/0` or `1**oo`  but limits from atleast one side exist.  For example, if `x=1`, then `(x**2-1)/(x-1)` directly valuates to `nan` or raises an error, although we can all be reasonably convinced that we can treat the actual value as `2` without problems.   `None`, on the other hand, while very similar, has special meaning in python usage that is more general than just `NaN`.  The concept of *not-a-number*, after all, was specifically created for numerical calculations that fail to evaluate (which to me is the equivalent of having been left undefined).

**Comment by zanzibar7** [comment]
`NaN` occurs naturally in the kinds of calculations associated with `Heaviside(0)`.  In particular, situations where direct evaluation gives `0/0` or `1**oo`  but limits from atleast one side exist.  For example, if `x=1`, then `(x**2-1)/(x-1)` directly valuates to `nan` or raises an error, although we can all be reasonably convinced that we can treat the actual value as `2` without problems.   `None`, on the other hand, while very similar, has special meaning in python usage that is more general than just `NaN`.  The concept of *not-a-number*, after all, was specifically created for numerical calculations that fail to evaluate (which to me is the equivalent of having been left undefined).

**Comment by zanzibar7** [comment]
Well, that's a clever idea -- I'd never seen `_sentinel` used  before, and will keep in mind for the future (I see something similar was used in `InverseMellinTransform` .  But it feels awkward.

I like the idea of a generalization to an variety of Heaviside functions with different handlings of `Heaviside(0)`.  But there too, the complications in `Function`'s implementation make me pause.

**Comment by cbm755** [comment]
1. One special thing about `None` is it survives `sympify()`.  I'm not sure this "fix" is worth pursuing further.  Do you think it is?

2. re: NaN: I don't disagree with any of that.  I will note that `Heaviside(0, nan) == nan` is not inconsistent with your logic.

3.  I agree re: sentinel; cool but probably a sign the code is too complicated :)  The other PR was *so* much easier to do so I'm glad others prefer it.

File: sympy/functions/special/delta_functions.py [hunk 1]
```diff
@@ -349,9 +349,29 @@ class Heaviside(Function):
...
 
     .. [*] Regarding to the value at 0, Mathematica defines ``H(0) = 1``,
-           but Maple uses ``H(0) = undefined``
+           but Maple uses ``H(0) = undefined``.  Different application areas
+           may have specific convensions.  For example, in control theory, it
+           is common practice to assume ``H(0) == 0`` to match the Laplace
+           transform of a DiracDelta distribution.
+
+    To specify the value of Heaviside at x=0, a second argument can be given.
+    Omit this 2nd argument or pass ``None`` to recover the default behavior.
```
Comment (sampadsaha5) [inline sympy/functions/special/delta_functions.py]: This seems a good idea. The idea of "None".

File: sympy/functions/special/delta_functions.py [hunk 2]
```diff
@@ -437,16 +463,23 @@ def eval(cls, arg):
...
+            return Heaviside(arg)
+        if arg.is_negative:
             return S.Zero
         elif arg.is_positive:
             return S.One
+        elif arg.is_zero:
+            if H0 is cls._sentinel:
+                return
+            H0 = sympify(H0)
+            return H0
```
Comment (sampadsaha5) [inline sympy/functions/special/delta_functions.py]: I think we should allow H0 == 1 and H0==1/2 only.

**Comment by cbm755** [comment]
We merged the better version of this, so I'll close this one.

@sampadsaha5 thanks for your comments!  Can you file a new issue for `H0 == 1 and H0 = 1/2 only`?  In particular, why do you think those are the only choices anyone would need?

**Pull request closed without merging**